### ISSUE TITLE
Library no longer compiles on platformio for ESP32

### ISSUE THREAD

**Issue by mbariola** [adafruit/Adafruit-PN532#69]
- Arduino board:  **INSERT ARDUINO BOARD NAME/TYPE HERE**

Esp32 Dev Module w/SSD1306 Screen

- Arduino IDE version (found in Arduino -> About Arduino menu):

PlarformIO for VS Code, Home v3.0.1 Core v4.1.0

- List the steps to reproduce the problem below (if possible attach a sketch or
  copy the sketch code in too): **LIST REPRO STEPS BELOW**

I have compiled this sketch successfully many times up to a couple month ago when I last used it. Then I tried recompiling and got these errors (and many warnings too, but the warnings have always been there - the errors forbidding me to compile is something new):

```
In file included from .pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.cpp:60:0:
.pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.h: In constructor 'Adafruit_PN532::Adafruit_PN532(uint8_t, uint8_t, uint8_t, uint8_t)':
.pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.h:202:29: error: 'Adafruit_PN532::_miso' will be initialized after [-Werror=reorder]   
   uint8_t _ss, _clk, _mosi, _miso;
                             ^
.pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.h:202:22: error:   'uint8_t Adafruit_PN532::_mosi' [-Werror=reorder]
   uint8_t _ss, _clk, _mosi, _miso;
                      ^
.pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.cpp:127:1: error:   when initialized here [-Werror=reorder]
 Adafruit_PN532::Adafruit_PN532(uint8_t clk, uint8_t miso, uint8_t mosi, uint8_t ss):
 ^
In file included from .pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.cpp:60:0:
.pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.h:202:22: error: 'Adafruit_PN532::_mosi' will be initialized after [-Werror=reorder]
   uint8_t _ss, _clk, _mosi, _miso;
                      ^
.pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.h:202:11: error:   'uint8_t Adafruit_PN532::_ss' [-Werror=reorder]
   uint8_t _ss, _clk, _mosi, _miso;
           ^
Compiling C:\Users\Massi\AppData\Local\Temp\pio-build\CROVE_RFID_Test-7a3db15896\esp32dev\lib142\BLE\BLEBeacon.cpp.o
.pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.cpp:127:1: error:   when initialized here [-Werror=reorder]
 Adafruit_PN532::Adafruit_PN532(uint8_t clk, uint8_t miso, uint8_t mosi, uint8_t ss):
 ^
In file included from .pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.cpp:60:0:
.pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.h: In constructor 'Adafruit_PN532::Adafruit_PN532(uint8_t, uint8_t)':
.pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.h:202:29: error: 'Adafruit_PN532::_miso' will be initialized after [-Werror=reorder]
   uint8_t _ss, _clk, _mosi, _miso;
                             ^
.pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.h:202:22: error:   'uint8_t Adafruit_PN532::_mosi' [-Werror=reorder]
   uint8_t _ss, _clk, _mosi, _miso;
                      ^
.pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.cpp:152:1: error:   when initialized here [-Werror=reorder]
 Adafruit_PN532::Adafruit_PN532(uint8_t irq, uint8_t reset):
 ^
In file included from .pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.cpp:60:0:
.pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.h:202:22: error: 'Adafruit_PN532::_mosi' will be initialized after [-Werror=reorder]
   uint8_t _ss, _clk, _mosi, _miso;
                      ^
.pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.h:202:11: error:   'uint8_t Adafruit_PN532::_ss' [-Werror=reorder]
   uint8_t _ss, _clk, _mosi, _miso;
           ^
.pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.cpp:152:1: error:   when initialized here [-Werror=reorder]
 Adafruit_PN532::Adafruit_PN532(uint8_t irq, uint8_t reset):
 ^
In file included from .pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.cpp:60:0:
.pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.h: In constructor 'Adafruit_PN532::Adafruit_PN532(uint8_t)':
.pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.h:202:29: error: 'Adafruit_PN532::_miso' will be initialized after [-Werror=reorder]
   uint8_t _ss, _clk, _mosi, _miso;
                             ^
.pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.h:202:22: error:   'uint8_t Adafruit_PN532::_mosi' [-Werror=reorder]
   uint8_t _ss, _clk, _mosi, _miso;
                      ^
.pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.cpp:173:1: error:   when initialized here [-Werror=reorder]
 Adafruit_PN532::Adafruit_PN532(uint8_t ss):
 ^
In file included from .pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.cpp:60:0:
.pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.h:202:22: error: 'Adafruit_PN532::_mosi' will be initialized after [-Werror=reorder]
   uint8_t _ss, _clk, _mosi, _miso;
                      ^
.pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.h:202:11: error:   'uint8_t Adafruit_PN532::_ss' [-Werror=reorder]
   uint8_t _ss, _clk, _mosi, _miso;
           ^
.pio\libdeps\esp32dev\Adafruit PN532_ID29\Adafruit_PN532.cpp:173:1: error:   when initialized here [-Werror=reorder]

```

The sketch I am running is your readntag203.pde to read NTAG203 / NTAG213 cards, adapted to display also the tag UID on the display. Here it is:

```

/**************************************************************************/
/*! 
    @file     readntag203.pde
    @author   KTOWN (Adafruit Industries)
    @license  BSD (see license.txt)

    This example will wait for any NTAG203 or NTAG213 card or tag,
    and will attempt to read from it.

    This is an example sketch for the Adafruit PN532 NFC/RFID breakout boards
    This library works with the Adafruit NFC breakout 
      ----> https://www.adafruit.com/products/364
 
    Check out the links above for our tutorials and wiring diagrams 
    These chips use SPI or I2C to communicate.

    Adafruit invests time and resources providing this open source code, 
    please support Adafruit and open-source hardware by purchasing 
    products from Adafruit!
*/
/**************************************************************************/
#include <Arduino.h>
#include <Wire.h>
#include <SPI.h>
#include <Adafruit_PN532.h>
#include <SSD1306Wire.h>

// If using the breakout with SPI, define the pins for SPI communication.
#define PN532_SS   5
#define PN532_SCK  18
#define PN532_MOSI 23
#define PN532_MISO 19 // conect MISO w/MISO, MOSI w/MOSI

// If using the breakout or shield with I2C, define just the pins connected
// to the IRQ and reset lines.  Use the values below (2, 3) for the shield!
#define PN532_IRQ   (22)
#define PN532_RESET (21)  // Not connected by default on the NFC Shield
#define OLEDRESETPIN 16
#define CODE "CROVE_ESP32_RFIDTest"

// Uncomment just _one_ line below depending on how your breakout or shield
// is connected to the Arduino:

// Use this line for a breakout with a software SPI connection (recommended):
Adafruit_PN532 nfc(PN532_SCK, PN532_MISO, PN532_MOSI, PN532_SS);

// Use this line for a breakout with a hardware SPI connection.  Note that
// the PN532 SCK, MOSI, and MISO pins need to be connected to the Arduino's
// hardware SPI SCK, MOSI, and MISO pins.  On an Arduino Uno these are
// SCK = 13, MOSI = 11, MISO = 12.  The SS line can be any digital IO pin.
//Adafruit_PN532 nfc(PN532_SS);

// Or use this line for a breakout or shield with an I2C connection:
//Adafruit_PN532 nfc(PN532_IRQ, PN532_RESET);

#if defined(ARDUINO_ARCH_SAMD)
// for Zero, output on USB Serial console, remove line below if using programming port to program the Zero!
// also change #define in Adafruit_PN532.cpp library file
   #define Serial SerialUSB
#endif

SSD1306Wire display(0x3c, 4, 15);

void initDisplay() {
  pinMode(OLEDRESETPIN, OUTPUT);   // OLED reset pin
  digitalWrite(OLEDRESETPIN, LOW); // set GPIO16 low to reset OLED
  delay(50);
  digitalWrite(OLEDRESETPIN, HIGH); // while OLED is running, must set GPIO16 in high
  delay(50);
  display.init();
  display.flipScreenVertically();

}
void startupMessage()
{
  #if defined(CODE)
  Serial.println(CODE);
  display.clear();
  display.setTextAlignment(TEXT_ALIGN_CENTER);
  display.setFont(ArialMT_Plain_10);
  display.drawStringMaxWidth(64, 24, 128, CODE);
  display.display();
  delay(2000);
  #endif
}

void setup(void) {
  #ifndef ESP8266
    while (!Serial); // for Leonardo/Micro/Zero
  #endif
  Serial.begin(115200);
  Serial.println("Hello!");
  nfc.begin();
  uint32_t versiondata = nfc.getFirmwareVersion();
  Serial.println("Finding Board...");
while(! versiondata) {
Serial.println("Didn't find PN53x board");
    delay(1000);
    versiondata = nfc.getFirmwareVersion();
  }
  // Got ok data, print it out!
  Serial.print("Found chip PN5"); Serial.println((versiondata>>24) & 0xFF, HEX); 
  Serial.print("Firmware ver. "); Serial.print((versiondata>>16) & 0xFF, DEC); 
  Serial.print('.'); Serial.println((versiondata>>8) & 0xFF, DEC);
  
  // configure board to read RFID tags
  nfc.SAMConfig();
  
  Serial.println("Waiting for an ISO14443A Card ...");
  initDisplay();
  startupMessage();
}

void loop(void) {
  uint8_t success;
  uint8_t uid[] = { 0, 0, 0, 0, 0, 0, 0 };  // Buffer to store the returned UID
  uint8_t uidLength;                        // Length of the UID (4 or 7 bytes depending on ISO14443A card type)
    
  // Wait for an NTAG203 card.  When one is found 'uid' will be populated with
  // the UID, and uidLength will indicate the size of the UUID (normally 7)
  success = nfc.readPassiveTargetID(PN532_MIFARE_ISO14443A, uid, &uidLength);
  
  if (success) {
    // Display some basic information about the card
    display.clear();
    display.setFont(ArialMT_Plain_16);
    display.setTextAlignment(TEXT_ALIGN_CENTER);
    display.drawStringMaxWidth(64, 0, 128, "ISO14443A card");
    display.display();

    Serial.println("Found an ISO14443A card");
    Serial.print("  UID Length: ");Serial.print(uidLength, DEC);Serial.println(" bytes");
    Serial.print("  UID Value: ");
    nfc.PrintHex(uid, uidLength);
    Serial.println("");
    
    if (uidLength == 7)
    {
      uint8_t data[32];
      
      // We probably have an NTAG2xx card (though it could be Ultralight as well)
    display.clear();
    display.setFont(ArialMT_Plain_16);
    display.setTextAlignment(TEXT_ALIGN_CENTER);
    display.drawStringMaxWidth(64, 0, 128, "NTAG2xx tag (7 byte UID)");
    display.display();
      Serial.println("Seems to be an NTAG2xx tag (7 byte UID)");	  
      
      // NTAG2x3 cards have 39*4 bytes of user pages (156 user bytes),
      // starting at page 4 ... larger cards just add pages to the end of
      // this range:
      
      // See: http://www.nxp.com/documents/short_data_sheet/NTAG203_SDS.pdf

      // TAG Type       PAGES   USER START    USER STOP
      // --------       -----   ----------    ---------
      // NTAG 203       42      4             39
      // NTAG 213       45      4             39
      // NTAG 215       135     4             129
      // NTAG 216       231     4             225      

      for (uint8_t i = 0; i < 42; i++) 
      {
        success = nfc.ntag2xx_ReadPage(i, data);
        
        // Display the current page number
        Serial.print("PAGE ");
        if (i < 10)
        {
          Serial.print("0");
          Serial.print(i);
        }
        else
        {
          Serial.print(i);
        }
        Serial.print(": ");

        // Display the results, depending on 'success'
        if (success) 
        {
          // Dump the page data
          nfc.PrintHexChar(data, 4);
        }
        else
        {
          Serial.println("Unable to read the requested page!");
        }
      }
    }
    else
    {
      Serial.println("This doesn't seem to be an NTAG203 tag (UUID length != 7 bytes)!");
    }
    
    // Wait a bit before trying again
    Serial.println("\n\nSend a character to scan another tag!");
/*    Serial.flush();
    while (!Serial.available());
    while (Serial.available()) {
    Serial.read();
    }*/
    Serial.flush();    
  }
}

```

this, FWIW, is my platformio.ini file:
```

; PlatformIO Project Configuration File
;
;   Build options: build flags, source filter
;   Upload options: custom upload port, speed and extra flags
;   Library options: dependencies, extra library storages
;   Advanced options: extra scripting
;
; Please visit documentation for the other options and examples
; https://docs.platformio.org/page/projectconf.html

[platformio]
build_dir = ${sysenv.TEMP}/pio-build/$PROJECT_HASH

[env:esp32dev]
platform = espressif32
board = esp32dev
framework = arduino
lib_deps =
  Wire
  Adafruit PN532
  ESP8266_SSD1306
  SPI
monitor_speed = 115200
board_build.f_cpu = 80000000L
```

I want to stress that the sketch compiled fine up till now (it is running on a board, I just wanted to change a line but it will no longer successfully compile). It seems a problem of compiler options here but am not sure...

Any help or pointer in the right direction very much appreciated as am currently stuck.

**Comment by mbariola** [comment]
As suggested by @manuelbl [here:](https://community.platformio.org/t/error-adafruit-pn532-miso-will-be-initialized-after-werror-reorder/11275/3)

changing line 202 in Adafruit_PN532.h to 
`  uint8_t _clk, _miso, _mosi, _ss;`

provides a band-aid to the problem. Still it might be of interest investigating why the problem occurs (or if it breaks other environments) and provide a more robust fix

**Comment by ladyada** [comment]
fixed in 1.2.0+
